File: bigtree/web/static/language/language.js
(() => {
  "use strict";

  const $ = (id) => document.getElementById(id);
  let statusCache = null;
  let channelCache = [];
  let builtInPrompt = "";

  function authKey(){
    try{
      const parentStorage = window.parent && window.parent.localStorage;
      const key = parentStorage ? parentStorage.getItem("bt_api_key") : "";
      if (key) return key;
    }catch(_err){}
    try{
      return window.localStorage.getItem("bt_api_key") || window.sessionStorage.getItem("bt_api_key") || "";
    }catch(_err){ return ""; }
  }

  async function request(path, options = {}){
    const opts = {...options};
    opts.headers = {...(options.headers || {})};
    const key = authKey();
    if (key){
      opts.headers["X-API-Key"] = key;
      opts.headers.Authorization = `Bearer ${key}`;
    }
    if (opts.body && !opts.headers["Content-Type"]) opts.headers["Content-Type"] = "application/json";
    const response = await fetch(path, opts);
    const text = await response.text();
    let payload = {};
    try{ payload = text ? JSON.parse(text) : {}; }
    catch(_err){ throw new Error(`Unexpected response (${response.status})`); }
    if (!response.ok || payload.ok === false) throw new Error(payload.error || `HTTP ${response.status}`);
    return payload;
  }

  function notice(message, kind = ""){
    const el = $("pageNotice");
    if (!message){ el.textContent = ""; el.className = "notice hidden"; return; }
    el.textContent = message;
    el.className = `notice${kind ? " " + kind : ""}`;
  }

  function statusLine(message, kind = ""){
    const el = $("configStatus");
    el.textContent = message;
    el.className = `status-line${kind ? " " + kind : ""}`;
  }

  function stamp(value){
    if (!value) return "—";
    const date = new Date(value);
    return Number.isNaN(date.getTime()) ? String(value) : date.toLocaleString();
  }

  function bytes(value){
    let n = Number(value || 0);
    if (!Number.isFinite(n) || n <= 0) return "0 B";
    const units = ["B","KB","MB","GB"];
    let i = 0;
    while (n >= 1024 && i < units.length - 1){ n /= 1024; i += 1; }
    return `${n.toFixed(i ? 1 : 0)} ${units[i]}`;
  }

  function escapeHtml(value){
    return String(value ?? "").replace(/[&<>'"]/g, (ch) => ({
      "&":"&amp;", "<":"&lt;", ">":"&gt;", "'":"&#39;", '"':"&quot;"
    })[ch]);
  }

  function providerLabel(id){
    const providers = statusCache?.providers || [];
    return providers.find((item) => item.id === id)?.label || id || "—";
  }

  function renderStatus(language){
    statusCache = language || {};
    const features = statusCache.features || {};
    const runtime = statusCache.runtime || {};
    const memory = statusCache.memory_stats || {};
    const context = statusCache.context || {};
    const reverence = statusCache.reverence || {};
    const storage = statusCache.storage || {};

    $("metricProvider").textContent = providerLabel(statusCache.provider);
    $("metricKey").textContent = `${statusCache.key_hint || "Not configured"} · ${statusCache.key_source || "unknown"} · ${statusCache.key_kind || "key"}`;
    $("metricModel").textContent = statusCache.model || "—";
    $("metricReasoning").textContent = `Reasoning ${statusCache.reasoning_mode || "—"}`;
    $("metricAudience").textContent = features.priest_chat ? "Enabled" : "Disabled";
    $("metricPriest").textContent = `Priest gate preserved · reverence ${reverence.enabled ? "on" : "off"}`;
    $("metricResult").textContent = runtime.last_error ? "Error" : (runtime.last_success_at ? "OK" : "Never");
    $("metricUsage").textContent = runtime.last_success_at
      ? `${runtime.input_tokens ?? "?"} in · ${runtime.output_tokens ?? "?"} out · ${runtime.last_latency_ms ?? "?"} ms`
      : "No successful request recorded";
    $("metricMemory").textContent = features.memory ? "Enabled" : "Disabled";
    $("metricMemoryDetail").textContent = `${memory.pinned ?? 0} pinned · ${memory.conversation ?? 0} conversation rows · ${bytes(memory.content_bytes)}`;

    $("configSource").textContent = `key: ${statusCache.key_source || "—"}`;
    $("provider").value = statusCache.provider || "openai";
    $("model").value = statusCache.model || "";
    $("temperature").value = statusCache.temperature ?? 0.7;
    $("maxTokens").value = statusCache.max_output_tokens ?? 400;
    $("reasoningMode").value = statusCache.reasoning_mode || "automatic";
    $("keyKind").value = statusCache.key_kind || "—";
    $("apiKey").value = "";
    $("apiKey").placeholder = statusCache.key_configured
      ? `Current: ${statusCache.key_hint} — leave blank to keep`
      : "Paste provider API key";
    $("providerHelp").textContent = statusCache.provider === "minimax"
      ? "MiniMax M3 supports adaptive/on/off reasoning. sk-cp Token Plan and sk-api pay-as-you-go keys are stored only in PostgreSQL."
      : "OpenAI remains available as a provider. The selected provider credential is stored only in PostgreSQL.";
    $("checkQuota").disabled = statusCache.provider !== "minimax";

    $("priestChat").checked = !!features.priest_chat;
    $("memoryEnabled").checked = !!features.memory;
    $("memoryTurns").value = features.memory_turns ?? 6;
    $("memoryRetention").value = features.memory_retention_days ?? 90;
    $("memoryRowCap").value = features.memory_global_row_cap ?? 5000;
    $("storageDetail").textContent = `${storage.persistent_backend || "PostgreSQL"} · ${memory.conversation ?? 0}/${features.memory_global_row_cap ?? 5000} conversation rows · ${bytes(memory.content_bytes)} text · no local-disk memory archive`;

    $("reverenceEnabled").checked = reverence.enabled !== false;
    $("requireAddress").checked = reverence.require_proper_address !== false;
    $("priestFamiliarity").checked = reverence.priest_familiarity !== false;
    $("emergencyOverride").checked = reverence.emergency_override !== false;
    $("reverenceStrictness").value = reverence.strictness || "moderate";
    $("casualPolicy").value = reverence.casual_policy || "correct_only";
    $("acceptedTitles").value = (reverence.accepted_titles || []).join("\n");
    $("reverenceState").textContent = reverence.enabled === false ? "disabled" : `${reverence.strictness || "moderate"} · ${reverence.casual_policy === "correct_then_answer" ? "correct + answer" : "correct first"}`;

    $("discordContext").checked = !!features.discord_context;
    builtInPrompt = context.default_system_prompt || "";
    $("systemPrompt").value = context.system_prompt || builtInPrompt;
    $("contextMode").textContent = context.uses_override ? "Custom context" : "Built-in context";
    $("logicList").innerHTML = (statusCache.logic || []).map((line) => `<li>${escapeHtml(line)}</li>`).join("");

    const summary = runtime.context_summary || {};
    $("ctxReverence").textContent = summary.reverence || "—";
    $("ctxPinned").textContent = summary.pinned_memories ?? "—";
    $("ctxHistory").textContent = summary.history_messages ?? "—";
    $("ctxDiscord").textContent = summary.discord_excerpts ?? "—";
    $("ctxAllowed").textContent = summary.knowledge_allowed == null ? "—" : (summary.knowledge_allowed ? "yes" : "no — ritual correction only");

    $("diagProvider").textContent = providerLabel(runtime.provider || statusCache.provider);
    $("diagModel").textContent = runtime.model || statusCache.model || "—";
    $("diagReasoning").textContent = runtime.reasoning_mode || statusCache.reasoning_mode || "—";
    $("diagAttempt").textContent = stamp(runtime.last_attempt_at);
    $("diagSuccess").textContent = stamp(runtime.last_success_at);
    $("diagLatency").textContent = runtime.last_latency_ms == null ? "—" : `${runtime.last_latency_ms} ms`;
    $("diagInput").textContent = runtime.input_tokens ?? "—";
    $("diagOutput").textContent = runtime.output_tokens ?? "—";
    $("diagRequest").textContent = runtime.last_request_id || "—";
    $("diagError").textContent = runtime.last_error || "—";

    renderContextChannels(features.discord_context_channel_ids || []);
  }

  function renderChannels(channels, intent){
    channelCache = Array.isArray(channels) ? channels : [];
    const search = $("searchChannel");
    search.innerHTML = `<option value="">All readable channels</option>`;
    for (const channel of channelCache){
      const option = document.createElement("option");
      option.value = channel.id;
      option.textContent = channel.category ? `${channel.category} / #${channel.name}` : `#${channel.name}`;
      search.appendChild(option);
    }
    $("discordIntent").textContent = intent ? "message content available" : "message-content intent off";
    renderContextChannels(statusCache?.features?.discord_context_channel_ids || []);
  }

  function renderContextChannels(selectedIds){
    const selected = new Set((selectedIds || []).map(String));
    const select = $("contextChannels");
    select.innerHTML = "";
    for (const channel of channelCache){
      const option = document.createElement("option");
      option.value = channel.id;
      option.textContent = channel.category ? `${channel.category} / #${channel.name}` : `#${channel.name}`;
      option.selected = selected.has(String(channel.id));
      select.appendChild(option);
    }
    if (!channelCache.length){
      const option = document.createElement("option");
      option.textContent = "No readable text channels loaded";
      option.disabled = true;
      select.appendChild(option);
    }
  }

  function renderMemories(memories){
    const list = $("memoryList");
    if (!memories || !memories.length){ list.textContent = "No memory has been stored yet."; return; }
    list.innerHTML = memories.map((memory) => {
      const scope = memory.scope_type === "user" ? `user ${memory.scope_id}` : "global";
      const pin = memory.pinned ? "pinned" : "recent";
      return `<article class="memory-row">
        <div class="memory-meta"><span>${escapeHtml(scope)}</span><span>·</span><span>${escapeHtml(memory.kind)}</span><span>·</span><span>${escapeHtml(memory.role)}</span><span>·</span><span>${pin}</span><span>·</span><span>${escapeHtml(stamp(memory.created_at))}</span></div>
        <div class="memory-text">${escapeHtml(memory.content)}</div>
        <div class="memory-actions"><button class="danger-ghost" data-delete-memory="${memory.id}" type="button">Delete</button></div>
      </article>`;
    }).join("");
  }

  function renderSearch(results){
    const host = $("searchResults");
    if (!results || !results.length){ host.textContent = "No matching recent messages found."; return; }
    host.innerHTML = results.map((row) => `<article class="search-result">
      <div class="search-meta"><span>#${escapeHtml(row.channel_name)}</span><span>·</span><span>${escapeHtml(row.author_name)}</span><span>·</span><span>${escapeHtml(stamp(row.timestamp))}</span><span>·</span><span>score ${Number(row.score || 0).toFixed(1)}</span></div>
      <div class="search-text">${escapeHtml(row.content)}</div>
      ${row.jump_url ? `<div class="memory-actions"><a href="${escapeHtml(row.jump_url)}" target="_blank" rel="noopener">Open in Discord ↗</a></div>` : ""}
    </article>`).join("");
  }

  async function loadStatus(){ const payload = await request("/admin/language/status"); renderStatus(payload.language || {}); }
  async function loadChannels(){ const payload = await request("/admin/language/discord/channels"); renderChannels(payload.channels || [], !!payload.message_content_intent); }
  async function loadMemories(){ const payload = await request("/admin/language/memories?limit=100"); renderMemories(payload.memories || []); }

  async function loadAll(){
    notice("");
    try{ await Promise.all([loadStatus(), loadChannels(), loadMemories()]); }
    catch(err){ notice(err.message || "Unable to load Language Services.", "err"); }
  }

  function reverencePayload(){
    return {
      enabled: $("reverenceEnabled").checked,
      strictness: $("reverenceStrictness").value,
      require_proper_address: $("requireAddress").checked,
      casual_policy: $("casualPolicy").value,
      priest_familiarity: $("priestFamiliarity").checked,
      emergency_override: $("emergencyOverride").checked,
      accepted_titles: $("acceptedTitles").value.split(/\r?\n/).map((line) => line.trim()).filter(Boolean),
    };
  }

  async function saveConfig(){
    const button = $("saveConfig");
    button.disabled = true;
    statusLine("Saving…");
    const selectedChannels = Array.from($("contextChannels").selectedOptions || []).map((option) => option.value).filter(Boolean);
    const payload = {
      provider: $("provider").value,
      model: $("model").value.trim(),
      temperature: Number($("temperature").value || 0.7),
      max_output_tokens: Number($("maxTokens").value || 400),
      reasoning_mode: $("reasoningMode").value,
      enable_priest_chat: $("priestChat").checked,
      memory_enabled: $("memoryEnabled").checked,
      memory_turns: Number($("memoryTurns").value || 6),
      memory_retention_days: Number($("memoryRetention").value || 90),
      memory_global_row_cap: Number($("memoryRowCap").value || 5000),
      discord_context_enabled: $("discordContext").checked,
      discord_context_channel_ids: selectedChannels,
      system_prompt: $("systemPrompt").value.trim(),
      reverence: reverencePayload(),
    };
    const key = $("apiKey").value.trim();
    if (key) payload.api_key = key;
    try{
      const response = await request("/admin/language/config", {method:"POST", body:JSON.stringify(payload)});
      renderStatus(response.language || {});
      statusLine("Saved. New Priest audiences use these settings immediately.", "ok");
    }catch(err){ statusLine(err.message || "Save failed.", "err"); }
    finally{ button.disabled = false; }
  }

  async function clearApiKey(){
    if (!window.confirm(`Clear the stored ${providerLabel($("provider").value)} key from PostgreSQL?`)) return;
    try{
      const response = await request("/admin/language/config", {method:"POST", body:JSON.stringify({provider:$("provider").value, clear_api_key:true})});
      renderStatus(response.language || {});
      statusLine("Stored provider key cleared.", "ok");
    }catch(err){ statusLine(err.message || "Unable to clear key.", "err"); }
  }

  async function testProvider(){
    const button = $("testProvider");
    button.disabled = true;
    notice("Sending one small provider request…");
    try{
      const response = await request("/admin/language/test", {method:"POST", body:"{}"});
      renderStatus(response.language || {});
      notice(`Provider reply: ${response.reply || "OK"}`);
    }catch(err){ notice(err.message || "Provider test failed.", "err"); try{ await loadStatus(); }catch(_err){} }
    finally{ button.disabled = false; }
  }

  async function checkQuota(){
    const button = $("checkQuota");
    button.disabled = true;
    $("quotaStatus").textContent = "Checking MiniMax Token Plan quota…";
    try{
      const response = await request("/admin/language/quota");
      if (!response.supported){
        $("quotaStatus").textContent = response.message || "Quota lookup unavailable for this key type.";
      }else{
        const q = response.quota || {};
        const compact = Object.entries(q).slice(0, 8).map(([k,v]) => `${k}: ${typeof v === "object" ? JSON.stringify(v) : v}`).join(" · ");
        $("quotaStatus").textContent = compact || "MiniMax returned quota status successfully.";
      }
    }catch(err){ $("quotaStatus").textContent = err.message || "Quota lookup failed."; }
    finally{ button.disabled = statusCache?.provider !== "minimax"; }
  }

  async function addMemory(){
    const content = $("memoryContent").value.trim();
    if (!content){ notice("Enter a memory note first.", "err"); return; }
    const scope = $("memoryScope").value;
    const scopeId = scope === "user" ? $("memoryUserId").value.trim() : "";
    if (scope === "user" && !/^\d+$/.test(scopeId)){ notice("A Discord user memory needs a numeric user ID.", "err"); return; }
    try{
      await request("/admin/language/memories", {method:"POST", body:JSON.stringify({content, scope_type:scope, scope_id:scopeId, kind:"note", pinned:true})});
      $("memoryContent").value = "";
      await Promise.all([loadMemories(), loadStatus()]);
      notice("Memory pinned.");
    }catch(err){ notice(err.message || "Unable to save memory.", "err"); }
  }

  async function deleteMemory(id){
    try{ await request(`/admin/language/memories/${encodeURIComponent(id)}`, {method:"DELETE"}); await Promise.all([loadMemories(), loadStatus()]); }
    catch(err){ notice(err.message || "Unable to delete memory.", "err"); }
  }

  async function clearConversations(){
    if (!window.confirm("Clear all unpinned Priest conversation memory? Pinned notes remain.")) return;
    try{
      const response = await request("/admin/language/memories/clear-conversations", {method:"POST", body:"{}"});
      await Promise.all([loadMemories(), loadStatus()]);
      notice(`Cleared ${response.deleted || 0} conversation memory rows.`);
    }catch(err){ notice(err.message || "Unable to clear conversation memory.", "err"); }
  }

  async function pruneMemory(){
    try{
      const response = await request("/admin/language/memories/prune", {method:"POST", body:"{}"});
      renderStatus(response.language || {});
      await loadMemories();
      notice(`Pruned ${response.deleted || 0} expired/excess conversation rows.`);
    }catch(err){ notice(err.message || "Unable to prune memory.", "err"); }
  }

  async function searchDiscord(){
    const query = $("searchQuery").value.trim();
    if (query.length < 2){ $("searchResults").textContent = "Enter at least two characters to search."; return; }
    const channel = $("searchChannel").value;
    const params = new URLSearchParams({q:query, limit:"30"});
    if (channel) params.set("channel_id", channel);
    $("searchResults").textContent = "Searching recent Discord history…";
    try{ const response = await request(`/admin/language/discord/search?${params.toString()}`); renderSearch(response.results || []); }
    catch(err){ $("searchResults").textContent = err.message || "Discord search failed."; }
  }

  $("provider").addEventListener("change", () => {
    const provider = $("provider").value;
    const current = $("model").value.trim();
    if (provider === "minimax" && (!current || /^gpt-/i.test(current))) $("model").value = "MiniMax-M3";
    if (provider === "openai" && (!current || /^MiniMax-/i.test(current))) $("model").value = "gpt-4o-mini";
    $("apiKey").value = "";
    $("keyKind").value = "save/select provider to resolve";
    $("checkQuota").disabled = provider !== "minimax";
    $("providerHelp").textContent = provider === "minimax"
      ? "MiniMax: use sk-cp for Token Plan or sk-api for pay-as-you-go. The key is persisted in PostgreSQL only."
      : "OpenAI project/API keys remain supported and are persisted in PostgreSQL only.";
  });
  $("refreshAll").addEventListener("click", loadAll);
  $("saveConfig").addEventListener("click", saveConfig);
  $("clearApiKey").addEventListener("click", clearApiKey);
  $("testProvider").addEventListener("click", testProvider);
  $("checkQuota").addEventListener("click", checkQuota);
  $("restoreContext").addEventListener("click", () => { $("systemPrompt").value = builtInPrompt; $("contextMode").textContent = "Built-in context (pending save)"; });
  $("memoryScope").addEventListener("change", () => { $("memoryUserWrap").classList.toggle("hidden", $("memoryScope").value !== "user"); });
  $("addMemory").addEventListener("click", addMemory);
  $("clearConversations").addEventListener("click", clearConversations);
  $("pruneMemory").addEventListener("click", pruneMemory);
  $("memoryList").addEventListener("click", (event) => {
    const button = event.target.closest("[data-delete-memory]");
    if (button) deleteMemory(button.dataset.deleteMemory);
  });
  $("searchDiscord").addEventListener("click", searchDiscord);
  $("searchQuery").addEventListener("keydown", (event) => { if (event.key === "Enter") searchDiscord(); });

  loadAll();
})();

File: bigtree/webmods/language.py
from __future__ import annotations

import asyncio
from typing import Any, Dict, List

from aiohttp import web

import bigtree
from bigtree.inc import ai, discord_knowledge, language_memory
from bigtree.inc.database import get_database
from bigtree.inc.webserver import frontend_route, get_server, route


def _bool(value: Any, default: bool = False) -> bool:
    if isinstance(value, bool):
        return value
    if value is None:
        return default
    text = str(value).strip().lower()
    if text in {"1", "true", "yes", "on", "enabled"}:
        return True
    if text in {"0", "false", "no", "off", "disabled"}:
        return False
    return default


def _int(value: Any, default: int, minimum: int, maximum: int) -> int:
    try:
        parsed = int(value)
    except Exception:
        parsed = default
    return max(minimum, min(parsed, maximum))


def _float(value: Any, default: float, minimum: float, maximum: float) -> float:
    try:
        parsed = float(value)
    except Exception:
        parsed = default
    return max(minimum, min(parsed, maximum))


def _channel_ids(value: Any) -> List[str]:
    if isinstance(value, list):
        items = value
    elif isinstance(value, str):
        items = value.split(",")
    else:
        items = []
    out: List[str] = []
    seen = set()
    for item in items:
        channel_id = str(item or "").strip()
        if not channel_id or not channel_id.isdigit() or channel_id in seen:
            continue
        seen.add(channel_id)
        out.append(channel_id)
        if len(out) >= 20:
            break
    return out


def _titles(value: Any) -> List[str]:
    if isinstance(value, str):
        items = value.splitlines()
    elif isinstance(value, list):
        items = value
    else:
        items = []
    out: List[str] = []
    seen = set()
    for item in items:
        title = str(item or "").strip()[:80]
        if not title or title.casefold() in seen:
            continue
        seen.add(title.casefold())
        out.append(title)
        if len(out) >= 24:
            break
    return out


def _render_language_page() -> str:
    server = get_server()
    if server is None:
        return "<h1>Language Services</h1>"
    return server.render_template("language.html", {})


async def _language_status() -> Dict[str, Any]:
    try:
        memories = await asyncio.to_thread(language_memory.memory_stats)
    except Exception as exc:
        memories = {"total": 0, "pinned": 0, "conversation": 0, "users": 0, "content_bytes": 0, "error": str(exc)}
    status = ai.get_language_status()
    status["memory_stats"] = memories
    return status


@frontend_route("GET", "/admin/language", allow_public=True)
async def language_services_page(_req: web.Request) -> web.Response:
    return web.Response(text=_render_language_page(), content_type="text/html")


@route("GET", "/admin/language/status", scopes=["admin:web"])
async def language_status(_req: web.Request) -> web.Response:
    return web.json_response({"ok": True, "language": await _language_status()})


@route("POST", "/admin/language/config", scopes=["admin:web"])
async def update_language_config(req: web.Request) -> web.Response:
    try:
        body = await req.json()
    except Exception:
        return web.json_response({"ok": False, "error": "invalid json"}, status=400)
    if not isinstance(body, dict):
        return web.json_response({"ok": False, "error": "body must be an object"}, status=400)

    provider = str(body.get("provider") or "openai").strip().lower()
    if provider not in ai.PROVIDERS:
        return web.json_response({"ok": False, "error": "provider must be openai or minimax"}, status=400)

    db = get_database()
    config = await asyncio.to_thread(db.get_system_config, "language") or {}
    config = dict(config)
    config["provider"] = provider

    provider_keys = config.get("provider_keys")
    if not isinstance(provider_keys, dict):
        provider_keys = {}
    provider_keys = dict(provider_keys)
    if _bool(body.get("clear_api_key"), False):
        provider_keys.pop(provider, None)
    else:
        api_key = str(body.get("api_key") or "").strip()
        if api_key:
            if len(api_key) < 16:
                return web.json_response({"ok": False, "error": "API key looks too short"}, status=400)
            provider_keys[provider] = api_key[:512]
    config["provider_keys"] = provider_keys

    if "model" in body:
        model = str(body.get("model") or "").strip()
        if not model:
            return web.json_response({"ok": False, "error": "model is required"}, status=400)
        provider_models = config.get("provider_models")
        if not isinstance(provider_models, dict):
            provider_models = {}
        provider_models = dict(provider_models)
        provider_models[provider] = model[:120]
        config["provider_models"] = provider_models

    if "temperature" in body:
        config["temperature"] = _float(body.get("temperature"), 0.7, 0.0, 2.0)
    if "max_output_tokens" in body:
        config["max_output_tokens"] = _int(body.get("max_output_tokens"), 1200, 64, 16384)
    if "reasoning_mode" in body:
        mode = str(body.get("reasoning_mode") or "automatic").strip().lower()
        if mode not in {"automatic", "disabled"}:
            return web.json_response({"ok": False, "error": "invalid reasoning mode"}, status=400)
        config["reasoning_mode"] = mode
    if "enable_priest_chat" in body:
        config["enable_priest_chat"] = _bool(body.get("enable_priest_chat"), True)
    if "memory_enabled" in body:
        config["memory_enabled"] = _bool(body.get("memory_enabled"), True)
    if "memory_turns" in body:
        config["memory_turns"] = _int(body.get("memory_turns"), 6, 1, 20)
    if "memory_retention_days" in body:
        config["memory_retention_days"] = _int(body.get("memory_retention_days"), 90, 1, 365)
    if "memory_global_row_cap" in body:
        config["memory_global_row_cap"] = _int(body.get("memory_global_row_cap"), 5000, 100, 20000)
    if "discord_context_enabled" in body:
        config["discord_context_enabled"] = _bool(body.get("discord_context_enabled"), False)
    if "discord_context_channel_ids" in body:
        config["discord_context_channel_ids"] = _channel_ids(body.get("discord_context_channel_ids"))
    if "system_prompt" in body:
        system_prompt = str(body.get("system_prompt") or "").strip()
        config["system_prompt"] = system_prompt[:16000]

    if "reverence" in body:
        incoming = body.get("reverence")
        if not isinstance(incoming, dict):
            return web.json_response({"ok": False, "error": "reverence must be an object"}, status=400)
        reverence = config.get("reverence")
        if not isinstance(reverence, dict):
            reverence = {}
        reverence = dict(reverence)
        if "enabled" in incoming:
            reverence["enabled"] = _bool(incoming.get("enabled"), True)
        if "strictness" in incoming:
            strictness = str(incoming.get("strictness") or "moderate").strip().lower()
            if strictness not in {"gentle", "moderate", "ceremonial"}:
                return web.json_response({"ok": False, "error": "invalid reverence strictness"}, status=400)
            reverence["strictness"] = strictness
        if "require_proper_address" in incoming:
            reverence["require_proper_address"] = _bool(incoming.get("require_proper_address"), True)
        if "casual_policy" in incoming:
            policy = str(incoming.get("casual_policy") or "correct_only").strip().lower()
            if policy not in {"correct_only", "correct_then_answer"}:
                return web.json_response({"ok": False, "error": "invalid casual address policy"}, status=400)
            reverence["casual_policy"] = policy
        if "priest_familiarity" in incoming:
            reverence["priest_familiarity"] = _bool(incoming.get("priest_familiarity"), True)
        if "emergency_override" in incoming:
            reverence["emergency_override"] = _bool(incoming.get("emergency_override"), True)
        if "accepted_titles" in incoming:
            titles = _titles(incoming.get("accepted_titles"))
            reverence["accepted_titles"] = titles or list(ai.REVERENCE_DEFAULT_TITLES)
        config["reverence"] = reverence

    await asyncio.to_thread(db.update_system_config, "language", config)
    ai.reset_client_cache()
    return web.json_response({"ok": True, "language": await _language_status()})


@route("POST", "/admin/language/test", scopes=["admin:web"])
async def test_language_provider(req: web.Request) -> web.Response:
    try:
        body = await req.json()
    except Exception:
        body = {}
    prompt = str((body or {}).get("prompt") or "Reply with a short confirmation that the roots are listening.").strip()
    try:
        reply = await ai.ask(user_id=0, prompt=prompt[:1000], persona="plain")
    except Exception as exc:
        return web.json_response(
            {"ok": False, "error": f"{type(exc).__name__}: {str(exc)[:500]}", "language": await _language_status()},
            status=502,
        )
    return web.json_response({"ok": True, "reply": reply, "language": await _language_status()})


@route("GET", "/admin/language/quota", scopes=["admin:web"])
async def language_provider_quota(_req: web.Request) -> web.Response:
    try:
        quota = await ai.get_provider_quota()
    except Exception as exc:
        return web.json_response({"ok": False, "error": f"{type(exc).__name__}: {str(exc)[:500]}"}, status=502)
    return web.json_response({"ok": True, **quota})


@route("GET", "/admin/language/memories", scopes=["admin:web"])
async def list_language_memories(req: web.Request) -> web.Response:
    scope_type = (req.query.get("scope_type") or "").strip().lower() or None
    scope_id = req.query.get("scope_id")
    kind = (req.query.get("kind") or "").strip().lower() or None
    limit = _int(req.query.get("limit"), 100, 1, 500)
    try:
        memories = await asyncio.to_thread(
            language_memory.list_memories,
            scope_type=scope_type,
            scope_id=scope_id,
            kind=kind,
            limit=limit,
        )
    except Exception as exc:
        return web.json_response({"ok": False, "error": str(exc)}, status=500)
    return web.json_response({"ok": True, "memories": memories, "count": len(memories)})


@route("POST", "/admin/language/memories", scopes=["admin:web"])
async def create_language_memory(req: web.Request) -> web.Response:
    try:
        body = await req.json()
    except Exception:
        return web.json_response({"ok": False, "error": "invalid json"}, status=400)
    try:
        memory = await asyncio.to_thread(
            language_memory.add_memory,
            content=str(body.get("content") or ""),
            scope_type=str(body.get("scope_type") or "global"),
            scope_id=str(body.get("scope_id") or ""),
            kind=str(body.get("kind") or "note"),
            role="system",
            source="operator",
            pinned=_bool(body.get("pinned"), True),
        )
    except ValueError as exc:
        return web.json_response({"ok": False, "error": str(exc)}, status=400)
    except Exception as exc:
        return web.json_response({"ok": False, "error": str(exc)}, status=500)
    return web.json_response({"ok": True, "memory": memory})


@route("DELETE", "/admin/language/memories/{memory_id}", scopes=["admin:web"])
async def delete_language_memory(req: web.Request) -> web.Response:
    try:
        memory_id = int(req.match_info.get("memory_id") or 0)
    except Exception:
        memory_id = 0
    if memory_id <= 0:
        return web.json_response({"ok": False, "error": "invalid memory id"}, status=400)
    deleted = await asyncio.to_thread(language_memory.delete_memory, memory_id)
    if not deleted:
        return web.json_response({"ok": False, "error": "memory not found"}, status=404)
    return web.json_response({"ok": True})


@route("POST", "/admin/language/memories/clear-conversations", scopes=["admin:web"])
async def clear_language_conversations(req: web.Request) -> web.Response:
    try:
        body = await req.json()
    except Exception:
        body = {}
    user_id = (body or {}).get("user_id")
    if user_id in (None, ""):
        parsed_user_id = None
    else:
        try:
            parsed_user_id = int(user_id)
        except Exception:
            return web.json_response({"ok": False, "error": "user_id must be an integer"}, status=400)
    deleted = await asyncio.to_thread(language_memory.clear_conversation_memory, parsed_user_id)
    return web.json_response({"ok": True, "deleted": deleted})


@route("POST", "/admin/language/memories/prune", scopes=["admin:web"])
async def prune_language_conversations(_req: web.Request) -> web.Response:
    cfg = ai.get_language_config()
    deleted = await asyncio.to_thread(
        language_memory.prune_conversation_memory,
        retention_days=int(cfg.get("memory_retention_days", 90)),
        global_row_cap=int(cfg.get("memory_global_row_cap", 5000)),
    )
    return web.json_response({"ok": True, "deleted": deleted, "language": await _language_status()})


@route("GET", "/admin/language/discord/channels", scopes=["admin:web"])
async def language_discord_channels(_req: web.Request) -> web.Response:
    bot = getattr(bigtree, "bot", None)
    if bot is None:
        return web.json_response({"ok": False, "error": "Discord bot is not ready"}, status=503)
    channels = discord_knowledge.list_readable_channels(bot)
    return web.json_response({
        "ok": True,
        "channels": channels,
        "message_content_intent": bool(getattr(getattr(bot, "intents", None), "message_content", False)),
    })


@route("GET", "/admin/language/discord/search", scopes=["admin:web"])
async def language_discord_search(req: web.Request) -> web.Response:
    bot = getattr(bigtree, "bot", None)
    if bot is None:
        return web.json_response({"ok": False, "error": "Discord bot is not ready"}, status=503)
    query = str(req.query.get("q") or "").strip()
    if len(query) < 2:
        return web.json_response({"ok": False, "error": "Search query must be at least 2 characters"}, status=400)
    channel_id = str(req.query.get("channel_id") or "").strip()
    limit = _int(req.query.get("limit"), 30, 1, 100)
    channel_ids = [channel_id] if channel_id else None
    rows = await discord_knowledge.search_discord(
        bot,
        query,
        channel_ids=channel_ids,
        limit=limit,
        per_channel=500 if channel_id else 120,
        max_channels=1 if channel_id else 20,
        include_bots=False,
    )
    return web.json_response({"ok": True, "query": query, "results": rows, "count": len(rows)})
